In [2]:
import pandas as pd
import numpy as np


In [8]:
df=pd.read_csv('price_crops.csv')

In [40]:

df.head()

District Name             Commodity Modal Price (Rs./Quintal)   Price Date  \
0        Mysore  Ragi (Finger Millet)                    3500.0  27 Oct 2023   
1        Mysore  Ragi (Finger Millet)                    3800.0  13 Dec 2023   
2        Mysore  Ragi (Finger Millet)                    2023.0  02 Jan 2023   
3        Mysore  Ragi (Finger Millet)                    2391.0  07 Mar 2023   
4        Mysore  Ragi (Finger Millet)                    2292.0  09 Jan 2023   

  date month  monthn  
0   27   Oct      10  
1   13   Dec      12  
2   02   Jan       1  
3   07   Mar       3  
4   09   Jan       1

In [41]:
df.tail()

District Name Commodity Modal Price (Rs./Quintal)   Price Date date  \
6756        Tumkur   Coconut                     11500  29 Jun 2023   29   
6757        Tumkur   Coconut                     12500  05 Dec 2023   05   
6758        Tumkur   Coconut                     11500  06 Jul 2023   06   
6759        Tumkur   Coconut                     11500  26 Jun 2023   26   
6760        Tumkur   Coconut                     12500  12 Jul 2023   12   

     month  monthn  
6756   Jun       6  
6757   Dec      12  
6758   Jul       7  
6759   Jun       6  
6760   Jul       7

In [9]:
df.drop(['Unnamed: 0','Sl no.','Market Name','Variety','Grade','Min Price (Rs./Quintal)','Max Price (Rs./Quintal)'],axis=1,inplace=True)

In [11]:
df[df.isnull().any(axis=1)]

District Name Commodity Modal Price (Rs./Quintal) Price Date
1266           NaN       NaN                       NaN        NaN
1267     Bangalore     Local               02 Mar 2023        NaN

In [12]:
df=df.dropna().reset_index(drop=True)

In [13]:
df[df.isnull().any(axis=1)]

Empty DataFrame
Columns: [District Name, Commodity, Modal Price (Rs./Quintal), Price Date]
Index: []

In [24]:
temp=df['Price Date'].str.split(' ')
temp

0       [27, Oct, 2023]
1       [13, Dec, 2023]
2       [02, Jan, 2023]
3       [07, Mar, 2023]
4       [09, Jan, 2023]
             ...       
6756    [29, Jun, 2023]
6757    [05, Dec, 2023]
6758    [06, Jul, 2023]
6759    [26, Jun, 2023]
6760    [12, Jul, 2023]
Name: Price Date, Length: 6761, dtype: object

In [27]:
date=[]
for i in range(0,len(temp)):
    date.append(temp[i][0])

In [34]:
month=[]
for i in range(0,len(temp)):
    month.append(temp[i][1])#month

In [33]:
#date

In [ ]:
df['date']=date

In [35]:
df['month']=month

In [38]:
fin=[]
for i in month:
    if(i=='Jan'):
         fin.append(1)
         continue
    if(i=='Feb'):
         fin.append(2)
         continue
    if(i=='Mar'):
         fin.append(3)
         continue
    if(i=='Apr'):
         fin.append(4)
         continue
    if(i=='May'):
         fin.append(5)
         continue
    if(i=='Jun'):
         fin.append(6)
         continue
    if(i=='Jul'):
         fin.append(7)
         continue
    if(i=='Aug'):
         fin.append(8)
         continue
    if(i=='Sep'):
         fin.append(9)
         continue
    if(i=='Oct'):
         fin.append(10)
         continue
    if(i=='Nov'):
         fin.append(11)
         continue
    if(i=='Dec'):
         fin.append(12)
         continue
    

In [39]:
df['monthn']=fin

In [42]:
# lets start encodimg

In [43]:
df.head()

District Name             Commodity Modal Price (Rs./Quintal)   Price Date  \
0        Mysore  Ragi (Finger Millet)                    3500.0  27 Oct 2023   
1        Mysore  Ragi (Finger Millet)                    3800.0  13 Dec 2023   
2        Mysore  Ragi (Finger Millet)                    2023.0  02 Jan 2023   
3        Mysore  Ragi (Finger Millet)                    2391.0  07 Mar 2023   
4        Mysore  Ragi (Finger Millet)                    2292.0  09 Jan 2023   

  date month  monthn  
0   27   Oct      10  
1   13   Dec      12  
2   02   Jan       1  
3   07   Mar       3  
4   09   Jan       1

In [44]:
data=df.drop(['Price Date','month'],axis=1)

In [45]:
data.head()

District Name             Commodity Modal Price (Rs./Quintal) date  monthn
0        Mysore  Ragi (Finger Millet)                    3500.0   27      10
1        Mysore  Ragi (Finger Millet)                    3800.0   13      12
2        Mysore  Ragi (Finger Millet)                    2023.0   02       1
3        Mysore  Ragi (Finger Millet)                    2391.0   07       3
4        Mysore  Ragi (Finger Millet)                    2292.0   09       1

In [57]:
data['day']

0       27
1       13
2        2
3        7
4        9
        ..
6756    29
6757     5
6758     6
6759    26
6760    12
Name: day, Length: 6761, dtype: int64

In [51]:
len(data['monthn'].value_counts() )

12

In [50]:
len(data['date'].value_counts() )

31

In [58]:
data.head()

District Name             Commodity Modal Price (Rs./Quintal) date  monthn  \
0        Mysore  Ragi (Finger Millet)                    3500.0   27      10   
1        Mysore  Ragi (Finger Millet)                    3800.0   13      12   
2        Mysore  Ragi (Finger Millet)                    2023.0   02       1   
3        Mysore  Ragi (Finger Millet)                    2391.0   07       3   
4        Mysore  Ragi (Finger Millet)                    2292.0   09       1   

   day  
0   27  
1   13  
2    2  
3    7  
4    9

In [54]:
# Convert 'day' and 'monthn' columns to numeric (if not already)
data['day'] = pd.to_numeric(data['date'], errors='raise')  # 'coerce' will turn invalid parsing into NaN
data['monthn'] = pd.to_numeric(data['monthn'], errors='raise')

In [97]:

data['price'] = pd.to_numeric(data['Modal Price (Rs./Quintal)'], errors='raise')


In [59]:
data.drop('date',axis=1,inplace=True)

In [61]:
data.head()

District Name             Commodity Modal Price (Rs./Quintal)  monthn  day
0        Mysore  Ragi (Finger Millet)                    3500.0      10   27
1        Mysore  Ragi (Finger Millet)                    3800.0      12   13
2        Mysore  Ragi (Finger Millet)                    2023.0       1    2
3        Mysore  Ragi (Finger Millet)                    2391.0       3    7
4        Mysore  Ragi (Finger Millet)                    2292.0       1    9

In [62]:
data['day_sin'] = np.sin(2 * np.pi * data['day'] / 31)
data['day_cos'] = np.cos(2 * np.pi * data['day'] / 31)


data['month_sin'] = np.sin(2 * np.pi * data['monthn'] / 12)
data['month_cos'] = np.cos(2 * np.pi * data['monthn'] / 12)

In [65]:
data.head()

District Name             Commodity Modal Price (Rs./Quintal)   day_sin  \
0        Mysore  Ragi (Finger Millet)                    3500.0 -0.724793   
1        Mysore  Ragi (Finger Millet)                    3800.0  0.485302   
2        Mysore  Ragi (Finger Millet)                    2023.0  0.394356   
3        Mysore  Ragi (Finger Millet)                    2391.0  0.988468   
4        Mysore  Ragi (Finger Millet)                    2292.0  0.968077   

    day_cos     month_sin     month_cos  
0  0.688967 -8.660254e-01  5.000000e-01  
1 -0.874347 -2.449294e-16  1.000000e+00  
2  0.918958  5.000000e-01  8.660254e-01  
3  0.151428  1.000000e+00  6.123234e-17  
4 -0.250653  5.000000e-01  8.660254e-01

In [64]:
data.drop(['day','monthn'],axis=1,inplace=True)

In [ ]:
import pandas as pd

# Example: Assuming 'district' and 'crop_name' columns in your dataset
# Apply one-hot encoding to 'district' and 'crop_name'
data = pd.get_dummies(data, columns=['District Name', 'Commodity'], drop_first=True)

# Check the updated DataFrame


KeyError: "None of [Index(['District Name', 'Commodity'], dtype='object')] are in the [columns]"

In [93]:
data.head()


Modal Price (Rs./Quintal)   day_sin   day_cos     month_sin     month_cos  \
0                    3500.0 -0.724793  0.688967 -8.660254e-01  5.000000e-01   
1                    3800.0  0.485302 -0.874347 -2.449294e-16  1.000000e+00   
2                    2023.0  0.394356  0.918958  5.000000e-01  8.660254e-01   
3                    2391.0  0.988468  0.151428  1.000000e+00  6.123234e-17   
4                    2292.0  0.968077 -0.250653  5.000000e-01  8.660254e-01   

   District Name_Chamrajnagar  District Name_Hassan  District Name_Mysore  \
0                       False                 False                  True   
1                       False                 False                  True   
2                       False                 False                  True   
3                       False                 False                  True   
4                       False                 False                  True   

   District Name_Tumkur  Commodity_Onion  Commodity_Ragi (Finger Millet)  \
0                 False            False                            True   
1                 False            False                            True   
2                 False            False                            True   
3                 False            False                            True   
4                 False            False                            True   

   Commodity_Rice  Commodity_Tomato  
0           False             False  
1           False             False  
2           False             False  
3           False             False  
4           False             False

In [70]:
data.isnull().sum()


Modal Price (Rs./Quintal)         0
day_sin                           0
day_cos                           0
month_sin                         0
month_cos                         0
District Name_Chamrajnagar        0
District Name_Hassan              0
District Name_Mysore              0
District Name_Tumkur              0
Commodity_Onion                   0
Commodity_Ragi (Finger Millet)    0
Commodity_Rice                    0
Commodity_Tomato                  0
dtype: int64

In [71]:
len(data)

6761

In [94]:
# Convert boolean columns to integers (0 or 1)
boolean_columns = [
    'District Name_Chamrajnagar', 'District Name_Hassan', 'District Name_Mysore',
    'District Name_Tumkur', 'Commodity_Onion', 'Commodity_Ragi (Finger Millet)',
    'Commodity_Rice', 'Commodity_Tomato'
]

for col in boolean_columns:
    data[col] = data[col].astype(int)


In [99]:
data.drop('Modal Price (Rs./Quintal)',axis=1,inplace=True)

In [100]:
data.dtypes

day_sin                           float64
day_cos                           float64
month_sin                         float64
month_cos                         float64
District Name_Chamrajnagar          int32
District Name_Hassan                int32
District Name_Mysore                int32
District Name_Tumkur                int32
Commodity_Onion                     int32
Commodity_Ragi (Finger Millet)      int32
Commodity_Rice                      int32
Commodity_Tomato                    int32
price                             float64
dtype: object

In [196]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor

# Features (X) and target (y)
X = data.drop(columns=['price'])  # All columns except price
y = data['price']  # Target variable

# Split the data into training and testing sets (80% training, 20% testing)
X_train, X_test, Y_train, Y_test = train_test_split(X, y, test_size=0.2, random_state=42)




In [102]:
X

day_sin   day_cos     month_sin     month_cos  \
0    -0.724793  0.688967 -8.660254e-01  5.000000e-01   
1     0.485302 -0.874347 -2.449294e-16  1.000000e+00   
2     0.394356  0.918958  5.000000e-01  8.660254e-01   
3     0.988468  0.151428  1.000000e+00  6.123234e-17   
4     0.968077 -0.250653  5.000000e-01  8.660254e-01   
...        ...       ...           ...           ...   
6756 -0.394356  0.918958  1.224647e-16 -1.000000e+00   
6757  0.848644  0.528964 -2.449294e-16  1.000000e+00   
6758  0.937752  0.347305 -5.000000e-01 -8.660254e-01   
6759 -0.848644  0.528964  1.224647e-16 -1.000000e+00   
6760  0.651372 -0.758758 -5.000000e-01 -8.660254e-01   

      District Name_Chamrajnagar  District Name_Hassan  District Name_Mysore  \
0                              0                     0                     1   
1                              0                     0                     1   
2                              0                     0                     1   
3                              0                     0                     1   
4                              0                     0                     1   
...                          ...                   ...                   ...   
6756                           0                     0                     0   
6757                           0                     0                     0   
6758                           0                     0                     0   
6759                           0                     0                     0   
6760                           0                     0                     0   

      District Name_Tumkur  Commodity_Onion  Commodity_Ragi (Finger Millet)  \
0                        0                0                               1   
1                        0                0                               1   
2                        0                0                               1   
3                        0                0                               1   
4                        0                0                               1   
...                    ...              ...                             ...   
6756                     1                0                               0   
6757                     1                0                               0   
6758                     1                0                               0   
6759                     1                0                               0   
6760                     1                0                               0   

      Commodity_Rice  Commodity_Tomato  
0                  0                 0  
1                  0                 0  
2                  0                 0  
3                  0                 0  
4                  0                 0  
...              ...               ...  
6756               0                 0  
6757               0                 0  
6758               0                 0  
6759               0                 0  
6760               0                 0  

[6761 rows x 12 columns]

In [103]:
y

0        3500.0
1        3800.0
2        2023.0
3        2391.0
4        2292.0
         ...   
6756    11500.0
6757    12500.0
6758    11500.0
6759    11500.0
6760    12500.0
Name: price, Length: 6761, dtype: float64

In [79]:
from sklearn.preprocessing import StandardScaler

In [119]:
scaler = StandardScaler()

# Scale the training data
X_train = scaler.fit_transform(X_train)

# Scale the test data using the same scaler
X_test= scaler.transform(X_test)

In [191]:
scaler

StandardScaler()

In [192]:
import pickle
with open('scaler.pkl', 'wb') as f:
    pickle.dump(scaler, f)

In [77]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression,Ridge,Lasso
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score,mean_absolute_error,mean_squared_error

In [90]:
# # function to evalute a model
# def evaluate_model(true,predicted):
#     mae=mean_absolute_error(true,predicted)
#     mse=mean_squared_error(true,predicted)
#     rmse=np.sqrt(mean_squared_error(true,predicted))
#     r2_sQuare=r2_score(true,predicted)
#     return mae,mse,r2_sQuare

def evaluate_model(true, predicted):
    mae = mean_absolute_error(true, predicted)
    mse = mean_squared_error(true, predicted)
    rmse = np.sqrt(mse)  # Calculating RMSE
    r2 = r2_score(true, predicted)  # Corrected variable name for R² score
    return mae, mse, rmse, r2


In [92]:
X.dtypes

day_sin                           float64
day_cos                           float64
month_sin                         float64
month_cos                         float64
District Name_Chamrajnagar           bool
District Name_Hassan                 bool
District Name_Mysore                 bool
District Name_Tumkur                 bool
Commodity_Onion                      bool
Commodity_Ragi (Finger Millet)       bool
Commodity_Rice                       bool
Commodity_Tomato                     bool
dtype: object

In [120]:
from sklearn.svm import SVR
from sklearn.ensemble import GradientBoostingRegressor, AdaBoostRegressor
from xgboost import XGBRegressor as xgb
#from sklearn.ensemble import GradientBoostingRegressor as xgb
import lightgbm as lgb
from sklearn.linear_model import ElasticNet

models = {
    "Linear Regression": LinearRegression(),
    "Lasso": Lasso(),
    "Ridge": Ridge(),
    "KNN": KNeighborsRegressor(),
    "Decision Tree": DecisionTreeRegressor(),
    "Random Forest": RandomForestRegressor(),
    "SVR": SVR(),
    "Gradient Boosting": GradientBoostingRegressor(),
    "XGBoost": xgb(),
    "AdaBoost": AdaBoostRegressor(),
    "LightGBM": lgb.LGBMRegressor(),
    "ElasticNet": ElasticNet()
}

# Iterate through models, train, predict, and evaluate
for name, model in models.items():
    model.fit(X_train, Y_train)  # Train model

    # Make predictions
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)
    
    model_train_mae, model_train_mse, model_train_rmse, model_train_r2 = evaluate_model(Y_train, y_train_pred)
    model_test_mae, model_test_mse, model_test_rmse, model_test_r2 = evaluate_model(Y_test, y_test_pred)

    print(name)
    print('Model performance for Training set')
    print("- Root Mean Squared Error: {:.4f}".format(model_train_rmse))
    print("- Mean Absolute Error: {:.4f}".format(model_train_mae))
    print("- R² Score: {:.4f}".format(model_train_r2))

    print('----------------------------------')

    print('Model performance for Test set')
    print("- Root Mean Squared Error: {:.4f}".format(model_test_rmse))
    print("- Mean Absolute Error: {:.4f}".format(model_test_mae))
    print("- R² Score: {:.4f}".format(model_test_r2))

    print('='*35)
    print('\n')


Linear Regression
Model performance for Training set
- Root Mean Squared Error: 2318.7477
- Mean Absolute Error: 1506.4085
- R² Score: 0.7803
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 2203.1740
- Mean Absolute Error: 1428.5009
- R² Score: 0.7702


Lasso
Model performance for Training set
- Root Mean Squared Error: 2318.7556
- Mean Absolute Error: 1506.0365
- R² Score: 0.7803
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 2203.0100
- Mean Absolute Error: 1428.0654
- R² Score: 0.7702


Ridge
Model performance for Training set
- Root Mean Squared Error: 2318.7513
- Mean Absolute Error: 1506.5163
- R² Score: 0.7803
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 2203.0436
- Mean Absolute Error: 1428.4905
- R² Score: 0.7702


KNN
Model performance for Training set
- Root Mean Squared Error: 1570.2597
- Mean Absolute Error: 816.2804
- R² Score: 0.8992


In [122]:
#1) gradient boosting  hyper tuning
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import GradientBoostingRegressor
from scipy.stats import uniform, randint

# Define the model
gb = GradientBoostingRegressor()

# Set the hyperparameter distribution
param_dist = {
    'n_estimators': randint(50, 500),
    'learning_rate': uniform(0.01, 0.2),
    'max_depth': randint(3, 10),
    'min_samples_split': randint(2, 10),
    'min_samples_leaf': randint(1, 10)
}

# RandomizedSearchCV for quicker search
random_search = RandomizedSearchCV(estimator=gb, param_distributions=param_dist, n_iter=100, cv=5, n_jobs=-1, scoring='neg_mean_squared_error', random_state=42)
random_search.fit(X_train, Y_train)

# Get the best parameters and score
print("Best parameters:", random_search.best_params_)
print("Best score:", random_search.best_score_)

# Evaluate on test set
best_gb = random_search.best_estimator_
test_score = best_gb.score(X_test, Y_test)
print(f"Test R² Score: {test_score:.4f}")


Best parameters: {'learning_rate': 0.010104075399063163, 'max_depth': 7, 'min_samples_leaf': 9, 'min_samples_split': 4, 'n_estimators': 408}
Best score: -3080738.974864681
Test R² Score: 0.8652


In [139]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import RandomizedSearchCV
import numpy as np

# Define the parameter grid
param_dist = {
    'n_estimators': [100, 200, 500, 724],
    'max_depth': [10, 20, 30, None],
    'min_samples_split': [2, 5, 10, 12],
    'min_samples_leaf': [1, 2, 4],
    'max_features': ['sqrt', 'log2', None],  # 'auto' is not valid anymore
}

# Initialize the model
rf = RandomForestRegressor()

# Set up the randomized search with 5-fold cross-validation
random_search = RandomizedSearchCV(estimator=rf, param_distributions=param_dist, 
                                   n_iter=100, cv=5, verbose=2, n_jobs=-1, random_state=42)

# Fit the randomized search to the training data
random_search.fit(X_train, Y_train)

# Print the best parameters and best score
print("Best parameters:", random_search.best_params_)
print("Best score:", random_search.best_score_)

# Test R² score on the test set
test_score = random_search.score(X_test, Y_test)
print("Test R² Score:", test_score)


Fitting 5 folds for each of 100 candidates, totalling 500 fits
Best parameters: {'n_estimators': 100, 'min_samples_split': 5, 'min_samples_leaf': 4, 'max_features': None, 'max_depth': 10}
Best score: 0.8765829809348439
Test R² Score: 0.8597434915235971


In [140]:
import pickle

# Save the model
with open('model.pkl', 'wb') as file:
    pickle.dump(random_search.best_estimator_, file)


In [135]:
import lightgbm as lgb
from scipy.stats import randint, uniform

# Define the model
lgb_model = lgb.LGBMRegressor()

# Set the hyperparameter distribution
param_dist = {
    'num_leaves': randint(31, 150),
    'max_depth': [10, 20, -1],
    'learning_rate': uniform(0.01, 0.2),
    'n_estimators': randint(50, 500),
    'min_child_samples': randint(5, 20)
}

# RandomizedSearchCV for quicker search
random_search = RandomizedSearchCV(estimator=lgb_model, param_distributions=param_dist, n_iter=100, cv=5, n_jobs=-1, scoring='neg_mean_squared_error', random_state=42)
random_search.fit(X_train, Y_train)

# Get the best parameters and score
print("Best parameters:", random_search.best_params_)
print("Best score:", random_search.best_score_)

# Evaluate on test set
best_lgb = random_search.best_estimator_
test_score = best_lgb.score(X_test, Y_test)
print(f"Test R² Score: {test_score:.4f}")


KeyboardInterrupt: 

1) random forst
   Fitting 5 folds for each of 100 candidates, totalling 500 fits
Best parameters: {'n_estimators': 200, 'min_samples_split': 10, 'min_samples_leaf': 4, 'max_features': None, 'max_depth': 10}
Best score: 0.8763229295840901
Test R² Score: 0.8582925267253119


2) gradient boost
    Best parameters: {'learning_rate': 0.010104075399063163, 'max_depth': 7, 'min_samples_leaf': 9, 'min_samples_split': 4, 'n_estimators': 408}
Best score: -3080738.974864681
Test R² Score: 0.8652

3) lightgbm
    


In [136]:
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor, StackingRegressor
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

# Assuming your data is ready (X_train, X_test, y_train, y_test)

# Define base models
rf = RandomForestRegressor(n_estimators=200, min_samples_split=10, min_samples_leaf=4, max_features=None, max_depth=10)
gb = GradientBoostingRegressor(learning_rate=0.0101, max_depth=7, min_samples_leaf=9, min_samples_split=4, n_estimators=408)

# Define meta-model (a simple Linear Regression for stacking)
meta_model = LinearRegression()

# Define the stacking model
stacked_model = StackingRegressor(estimators=[('rf', rf), ('gb', gb)], final_estimator=meta_model)

# Train the stacking model
stacked_model.fit(X_train, Y_train)

# Make predictions
train_preds = stacked_model.predict(X_train)
test_preds = stacked_model.predict(X_test)

# Evaluate the model
train_r2 = r2_score(Y_train, train_preds)
test_r2 = r2_score(Y_test, test_preds)

print(f"Train R² Score: {train_r2:.4f}")
print(f"Test R² Score: {test_r2:.4f}")


Train R² Score: 0.8968
Test R² Score: 0.8628


In [138]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import StackingRegressor
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
import numpy as np

# Define base models
base_models = [
    ('rf', RandomForestRegressor(n_estimators=200, random_state=42)),
    ('gbr', GradientBoostingRegressor(n_estimators=200, random_state=42)),
    ('svr', SVR())
]

# Meta-model: Linear Regression
meta_model = LinearRegression()

# Stacking model
stacking_model = StackingRegressor(estimators=base_models, final_estimator=meta_model)

# Hyperparameter distribution for RandomForest and GradientBoosting
param_dist = {
    'rf__max_depth': [10, 20, 30, None],
    'rf__min_samples_split': [2, 5, 10],
    'gbr__learning_rate': [0.01, 0.1, 0.05],
    'gbr__max_depth': [3, 5, 7, 10],
    'gbr__min_samples_split': [2, 5, 10],
    'svr__C': [0.1, 1, 10, 100],
    'svr__kernel': ['linear', 'rbf'],
    'svr__epsilon': [0.01, 0.1, 0.5]
}

# RandomizedSearchCV with 5-fold cross-validation
random_search = RandomizedSearchCV(stacking_model, param_distributions=param_dist, n_iter=100, cv=5, scoring='r2', verbose=1, random_state=42)

# Fit model
random_search.fit(X_train, Y_train)

# Best parameters and score
print("Best parameters found: ", random_search.best_params_)
print("Best score found: ", random_search.best_score_)

# Test R² score
test_score = random_search.score(X_test, Y_test)
print(f'Test R² Score: {test_score}')


Fitting 5 folds for each of 100 candidates, totalling 500 fits


KeyboardInterrupt: 

In [189]:
X.iloc[0]

day_sin                          -0.724793
day_cos                           0.688967
month_sin                        -0.866025
month_cos                         0.500000
District Name_Chamrajnagar        0.000000
District Name_Hassan              0.000000
District Name_Mysore              1.000000
District Name_Tumkur              0.000000
Commodity_Onion                   0.000000
Commodity_Ragi (Finger Millet)    1.000000
Commodity_Rice                    0.000000
Commodity_Tomato                  0.000000
Name: 0, dtype: float64

In [173]:
df['Commodity'].value_counts()

Commodity
Tomato                  2566
Onion                   1609
Coconut                 1018
Rice                     860
Ragi (Finger Millet)     708
Name: count, dtype: int64

In [188]:
df[df['Commodity']=='Ragi (Finger Millet)'].head()
df.iloc[0]


District Name                              Mysore
Commodity                    Ragi (Finger Millet)
Modal Price (Rs./Quintal)                  3500.0
Price Date                            27 Oct 2023
date                                           27
month                                         Oct
monthn                                         10
Name: 0, dtype: object

In [ ]:
# ragi ==> [0,1,0,0]
#  tomato =[0,0,0,1]
# onion [1,0,0,0]
# rice =[0,0,1,0]
# ccocunut [0,0,0,0]


In [ ]:
# Mysore= [0,0,1,0]
# bangalore= [0,0,0,0]
# hassan =[0,1,0,0]
# tumkur =[0,0,0,1]
# chamr =[1,0,0,0]

In [190]:
X.head()

day_sin   day_cos     month_sin     month_cos  District Name_Chamrajnagar  \
0 -0.724793  0.688967 -8.660254e-01  5.000000e-01                           0   
1  0.485302 -0.874347 -2.449294e-16  1.000000e+00                           0   
2  0.394356  0.918958  5.000000e-01  8.660254e-01                           0   
3  0.988468  0.151428  1.000000e+00  6.123234e-17                           0   
4  0.968077 -0.250653  5.000000e-01  8.660254e-01                           0   

   District Name_Hassan  District Name_Mysore  District Name_Tumkur  \
0                     0                     1                     0   
1                     0                     1                     0   
2                     0                     1                     0   
3                     0                     1                     0   
4                     0                     1                     0   

   Commodity_Onion  Commodity_Ragi (Finger Millet)  Commodity_Rice  \
0                0                               1               0   
1                0                               1               0   
2                0                               1               0   
3                0                               1               0   
4                0                               1               0   

   Commodity_Tomato  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0

In [193]:
with open('scaler.pkl', 'rb') as f:
    loaded_scaler = pickle.load(f)

with open('model.pkl', 'rb') as f:
    model = pickle.load(f)


In [197]:
from sklearn.svm import SVR
from sklearn.ensemble import GradientBoostingRegressor, AdaBoostRegressor
from xgboost import XGBRegressor as xgb
#from sklearn.ensemble import GradientBoostingRegressor as xgb
import lightgbm as lgb
from sklearn.linear_model import ElasticNet


X_train=loaded_scaler.transform(X_train)
X_test=loaded_scaler.transform(X_test)

# Make predictions
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

model_train_mae, model_train_mse, model_train_rmse, model_train_r2 = evaluate_model(Y_train, y_train_pred)
model_test_mae, model_test_mse, model_test_rmse, model_test_r2 = evaluate_model(Y_test, y_test_pred)


print('Model performance for Training set')
print("- Root Mean Squared Error: {:.4f}".format(model_train_rmse))
print("- Mean Absolute Error: {:.4f}".format(model_train_mae))
print("- R² Score: {:.4f}".format(model_train_r2))

print('----------------------------------')

print('Model performance for Test set')
print("- Root Mean Squared Error: {:.4f}".format(model_test_rmse))
print("- Mean Absolute Error: {:.4f}".format(model_test_mae))
print("- R² Score: {:.4f}".format(model_test_r2))

print('='*35)
print('\n')


Model performance for Training set
- Root Mean Squared Error: 1570.9219
- Mean Absolute Error: 812.6681
- R² Score: 0.8992
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 1721.0932
- Mean Absolute Error: 880.6930
- R² Score: 0.8597


